pkg/tsdb/sqleng/resample.go

```diff
@@ -0,0 +1,128 @@
+package sqleng
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+func getRowFillValues(f *data.Frame, tsSchema data.TimeSeriesSchema, currentTime time.Time,
+	fillMissing *data.FillMissing, intermediateRows []int, lastSeenRowIdx int) []interface{} {
+	vals := make([]interface{}, 0, len(f.Fields))
+	for i, field := range f.Fields {
+		// if the current field is the time index of the series
+		// set the new value to be added to the new timestamp
+		if i == tsSchema.TimeIndex {
+			switch f.Fields[tsSchema.TimeIndex].Type() {
+			case data.FieldTypeTime:
+				vals = append(vals, currentTime)
+			default:
+				vals = append(vals, &currentTime)
+			}
+			continue
+		}
+
+		isValueField := false
+		for _, idx := range tsSchema.ValueIndices {
+			if i == idx {
+				isValueField = true
+				break
+			}
+		}
+
+		// if the current field is value Field
+		// set the new value to the last seen field value (if such exists)
+		// otherwise set the appropriate value according to the fillMissing mode
+		// if the current field is string field)
+		// set the new value to be added to the last seen value (if such exists)
+		// if the Frame is wide then there should not be any string fields
+		var newVal interface{}
+		if isValueField {
+			if len(intermediateRows) > 0 {
+				// instead of setting the last seen
+				// we could set avg, sum, min or max
+				// of the intermediate values for each field
+				newVal = f.At(i, intermediateRows[len(intermediateRows)-1])
+			} else {
+				val, err := data.GetMissing(fillMissing, field, lastSeenRowIdx)
+				if err == nil {
+					newVal = val
+				}
+			}
+		} else if lastSeenRowIdx >= 0 {
+			newVal = f.At(i, lastSeenRowIdx)
+		}
+		vals = append(vals, newVal)
+	}
+	return vals
+}
+
```

Comment:
It's still not crystal clear to me what this and the `getRowFillValues` function does. Could a comment that describes that be added?

Reply:
I pushed some updates. Not sure they make it easier to understand, but will ask Ying to review it.